### ISSUE TITLE
Global environment variables

### ISSUE THREAD

**Issue by magcius** [mesosphere/marathon#2923]
Might be a good or a bad idea depending on who you ask.

In our situation, we have a dev cluster and a prod cluster, with different external resources (e.g. S3 buckets). We use a shared naming scheme to build a bucket name from a cluster ID, but this requires that we go and edit every single task to match the new cluster name. It would be nice if we had a way to set global environment variables for all tasks like this.

Ideally, it would be from a file or command line parameter, so we could share the same Marathon configuration between clusters types -- having our startup scripts retrieve a tag from AWS would make everything automatic.

**Comment by aquamatthias** [comment]
@magcius thanks for this suggestion.
If I understand correctly, your goal is:
- have one configuration file that is passed to Marathon
- all properties are automatically available in all tasks as environment variables?

One suggestion: did you thought about defining the config file in the uri section of the application definition, so the config file will be available for all tasks?

**Comment by magcius** [comment]
That would require us to add in resources (URIs) for every single application definition, no? In addition, we'd have to set up some sort of stable service to fetch these URIs from, which would be cluster-global. And we'd also have to have some sort of patch to all our Docker containers to apply this URI as an envvar hook.

**Comment by aquamatthias** [comment]
@magcius Yes that would be the consequence. It looks like I understood your enhancement proposal correct.

**Comment by WalterDalton** [comment]
In addition, could there be way for Marathon to load the global environment variables from a file?  Like if there was a .sh script specified, it would take all VARs exported in that file and make then global for all tasks. That would allow a marathon instance to load a common file (of common variables), and allow other non-marathon frameworks to use the same file, with the same variables.

**Comment by magcius** [comment]
I would prefer if it wasn't a .sh file, simply because those are hard to parse if they contain subshells or other commands. The systemd `EnvironmentFile=` format would be perfect.

**Comment by llowrey** [comment]
It would be good if the implementation would allow for tasks to specify one or more env var files and supply the union to the task. Some of our vars contain security credentials that we don't want all tasks to see but we also have common vars that are safe for all tasks to see.

So, perhaps there could be three layers, all merged last-one-wins:

1. Global cluster-wide file

**Comment by magcius** [comment]
For reference, we just mounted a .ini as a read-only volume, and have been using that from app code. With a bit more effort, you could add a bash wrapper to all your Docker containers to source it. We actually preferred having app code reading vars from a file rather than envvars.

**Comment by jdef** [comment]
the marathon plugins API provides an interface that allows a plugin to inject things like envvars into the Mesos task definition at launch time. sounds like this use case could easily be met by a plugin impl.

**Comment by meichstedt** [comment]
Note: This issue has been migrated to https://jira.mesosphere.com/browse/MARATHON-2971. For more information see https://groups.google.com/forum/#!topic/marathon-framework/khtvf-ifnp8.

**Comment by meichstedt** [comment]
Note: This issue has been migrated to https://jira.mesosphere.com/browse/MARATHON-2971. For more information see https://groups.google.com/forum/#!topic/marathon-framework/khtvf-ifnp8.
